diff --git a/build/build.sh b/build/build.sh
index 5a6e5e7..7d4b095 100755
--- a/build/build.sh
+++ b/build/build.sh
@@ -43,6 +43,8 @@ ldflags="
   -X ${repo_path}/version.BuildDate${ldseparator}${build_date}
   -X ${repo_path}/version.GoVersion${ldseparator}${go_version}"
 
+ldflags="${ldflags} -extldflags '-static'"
+
 echo " >   cadvisor"
 godep go build -ldflags "${ldflags}" -o cadvisor${ext} ${repo_path}
 
