From aa7b93d65b6549ef7018d6106d4a4df632f02fba Mon Sep 17 00:00:00 2001
From: Thomas Orozco <thomas@orozco.fr>
Date: Wed, 6 Apr 2016 19:03:34 +0200
Subject: [PATCH] Report IO metrics to InfluxDB

---
 storage/influxdb/influxdb.go | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/storage/influxdb/influxdb.go b/storage/influxdb/influxdb.go
index 5e80e9f..1c245ab 100644
--- a/storage/influxdb/influxdb.go
+++ b/storage/influxdb/influxdb.go
@@ -73,6 +73,10 @@ const (
 	serFsLimit string = "fs_limit"
 	// Filesystem usage.
 	serFsUsage string = "fs_usage"
+	// Serviced IO bytes
+	serIoBytes string = "io_bytes"
+	// Serviced IO operations
+	serIoOps string = "io_ops"
 )
 
 func new() (storage.StorageDriver, error) {
@@ -201,6 +205,22 @@ func (self *influxdbStorage) containerStatsToPoints(
 	// RSS
 	points = append(points, makePoint(serMemoryRSS, stats.Memory.RSS))
 
+	// IO stats
+	var readBytes, writeBytes, readOps, writeOps uint64 = 0, 0, 0, 0
+
+	for _, diskStats := range stats.DiskIo.IoServiceBytes {
+		readBytes += diskStats.Stats["Read"]
+		writeBytes += diskStats.Stats["Write"]
+	}
+
+	for _, diskStats := range stats.DiskIo.IoServiced {
+		readOps += diskStats.Stats["Read"]
+		writeOps += diskStats.Stats["Write"]
+	}
+
+	points = append(points, makePoint(serIoBytes, readBytes+writeBytes))
+	points = append(points, makePoint(serIoOps, readOps+writeOps))
+
 	// Network Stats
 	points = append(points, makePoint(serRxBytes, stats.Network.RxBytes))
 	points = append(points, makePoint(serRxErrors, stats.Network.RxErrors))
-- 
2.7.4

