From aee69eaa750a82ac45d4f76ad6ba77dc0c4c9224 Mon Sep 17 00:00:00 2001
From: Thomas Orozco <thomas@orozco.fr>
Date: Mon, 11 Apr 2016 18:33:13 +0200
Subject: [PATCH] Support ecryptfs volumes

---
 fs/partition_cache.go | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/fs/partition_cache.go b/fs/partition_cache.go
index 000cf65..fc76a2a 100644
--- a/fs/partition_cache.go
+++ b/fs/partition_cache.go
@@ -77,9 +77,10 @@ func (self *RealPartitionCache) Refresh() error {
 
 	supportedFsType := map[string]bool{
 		// all ext systems are checked through prefix.
-		"btrfs": true,
-		"xfs":   true,
-		"zfs":   true,
+		"btrfs":    true,
+		"xfs":      true,
+		"zfs":      true,
+		"ecryptfs": true,
 	}
 
 	mounts, err := self.mountInfo.GetMounts()
-- 
2.7.4

